### PR TITLE
[QUIC] Basic Streams

### PR DESCRIPTION
**Pull request by Mr3zee**
**Subsystem**
ktor-network

**Solution**
Added basic implementations of stream management for QUIC.

### CODE REVIEW

File: ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/streams/QUICStreamChannel.kt [hunk 1]
```diff
@@ -7,23 +7,14 @@ package io.ktor.network.quic.streams
...
-    public val outgoing: SendChannel<QUICStream>
-
-    /**
-     * Send QUIC stream.
+     * Create new QUIC stream.
      */
-    public suspend fun send(stream: QUICStream) {
-        outgoing.send(stream)
-    }
+    public suspend fun createStream(): QUICStream
```
Comment (whyoleg) [inline ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/streams/QUICStreamChannel.kt:17]: there should be possibility to open both unidirectional and bidirectional streams

File: ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/streams/QUICStream.kt [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+package io.ktor.network.quic.streams
+
+import io.ktor.network.quic.bytes.*
+import io.ktor.utils.io.bits.*
+import io.ktor.utils.io.core.*
+
+public interface QUICStream {
+    public val streamId: Long
+
+    public val output: Output
```
Comment (whyoleg) [inline ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/streams/QUICStream.kt:14]: may be it makes more sense to use Byte[Read/Write]Channel for I/O ?

File: ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/connections/QUICConnection_v1.kt [hunk 3]
```diff
@@ -726,6 +762,89 @@ internal class QUICConnection_v1(
         private val logger = logger()
     }
 
+    private inner class StreamManager {
+        private val receiveStates = hashMapOf<Long, ReceiveStreamState>()
```
Comment (whyoleg) [inline ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/connections/QUICConnection_v1.kt:766]: note: Long will be boxed (not really an issue, just pointing)

also other note about synchronisation of this states, I mean, thread-safety
I don't see directly in code, if those can be used from multiple threads, so just pointing for additional attention.

File: ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/streams/QUICStreamChannel.kt (hunk 1 above)
Comment (Mr3zee) [reply ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/streams/QUICStreamChannel.kt:17]: Yes, for now it is only bidirectional to make PoC implementation (in current state in works for basic cases from the start to the end). I plan to follow the spec more precisely in next iterations, including support for both uni- and bi- directional streams

File: ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/connections/QUICConnection_v1.kt (hunk 3 above)
Comment (Mr3zee) [reply ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/connections/QUICConnection_v1.kt:766]: Yep, for now the goal was to make it work, it is definitely not the best approach. 

Multiple threads should not work with the same connection (at least for now)

File: ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/streams/QUICStream.kt (hunk 2 above)
Comment (Mr3zee) [reply ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/streams/QUICStream.kt:14]: Hm, need to check

File: ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/connections/QUICConnection_v1.kt [hunk 4]
```diff
@@ -726,6 +762,89 @@ internal class QUICConnection_v1(
...
+
+        private val streams = hashMapOf<Long, QUICStreamImpl>()
+
+        private val sendChannel: Channel<StreamFrame> = Channel(Channel.UNLIMITED)
+
+        init {
+            // todo close
+            CoroutineScope(Dispatchers.Default).launch {
+                while (isActive) {
```
Comment (e5l) [inline ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/connections/QUICConnection_v1.kt:775]: Could you tell me why it's needed?

File: ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/connections/QUICConnection_v1.kt [hunk 5]
```diff
@@ -726,6 +762,89 @@ internal class QUICConnection_v1(
...
+                )
+
+                receiveStates[streamId] = state
+            }
+
+            receiveStates[streamId]!!.receive(streamData, offset, fin)
+        }
+
+        private suspend fun onDataReceived(streamId: Long, dataChunk: ByteReadPacket) {
+            if (!streams.containsKey(streamId)) {
```
Comment (e5l) [inline ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/connections/QUICConnection_v1.kt:820]: invert and early return

File: ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/sockets/QUICSocketBase.kt [hunk 6]
```diff
@@ -27,7 +27,7 @@ internal abstract class QUICSocketBase(
 
     init {
         CoroutineScope(Dispatchers.Default).launch {
-            while (true) {
+            while (isActive) {
                 try {
                     receiveAndProcessDatagram()
                 } catch (e: Exception) {
```
Comment (e5l) [inline ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/sockets/QUICSocketBase.kt:33]: cause

**Review by e5l — CHANGES_REQUESTED** [review]
Please explain why `isActive` is needed. It also would be great to see some tests at this stage

File: ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/streams/QUICStream.kt (hunk 2 above)
Comment (Mr3zee) [reply ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/streams/QUICStream.kt:14]: I didn't find a quick way to use these interfaces here, and I need to finish my thesis by Monday, so I'll keep as it is for now, but I've added a TODO to revisit this place when I continue the work. Thank you the point though!

File: ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/connections/QUICConnection_v1.kt (hunk 4 above)
Comment (Mr3zee) [reply ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/connections/QUICConnection_v1.kt:775]: From the `isActive` doc:

Check this property in long-running computation loops to support cancellation:
```kotlin
while (isActive) {
     // do some computation
}
```

Comment (Mr3zee) [reply ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/connections/QUICConnection_v1.kt:775]: I know that I don't close it though :) But is does no harm here, and may prevent future me from forgetting to add it when I support cancelation.

File: ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/connections/QUICConnection_v1.kt (hunk 5 above)
Comment (Mr3zee) [reply ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/connections/QUICConnection_v1.kt:820]: There is a statement after `if` that should be executed, so I don't think it early return would improve the code here

**Pull request merged**